neqo-transport/src/pmtud.rs

```diff
@@ -0,0 +1,462 @@
+        // The packet may include ACK-eliciting data already, but rather than check for that, it
+        // seems OK to burn one byte here to simply include a PING.
+        builder.encode_varint(FRAME_TYPE_PING);
+        stats.ping += 1;
+        stats.all += 1;
+
+        self.probe_state = Probe::Prepared;
+        qtrace!(
+            "PMTUD probe of size {} prepared",
+            self.search_table[self.probe_index]
+        );
+    }
+
+    /// Records that a PMTUD probe has been sent.
+    pub fn probe_sent(&mut self, stats: &mut Stats) {
+        self.probe_state = Probe::Sent;
+        self.probe_count += 1;
+        stats.pmtud_tx += 1;
+        qdebug!(
+            "PMTUD probe of size {} sent, count {}",
+            self.search_table[self.probe_index],
+            self.probe_count
+        );
+    }
+
+    /// Returns true if the packet is a PMTUD probe.
+    #[must_use]
+    pub fn is_pmtud_probe(&self, p: &SentPacket) -> bool {
+        self.probe_state == Probe::Sent && p.len() == self.probe_size()
+    }
+
+    /// Count the PMTUD probes included in `pkts`.
+    fn count_pmtud_probes(&self, pkts: &[SentPacket]) -> usize {
+        pkts.iter().filter(|p| self.is_pmtud_probe(p)).count()
+    }
+
+    /// Checks whether a PMTUD probe has been acknowledged, and if so, updates the PMTUD state.
+    /// May also initiate a new probe process for a larger MTU.
+    pub fn on_packets_acked(&mut self, acked_pkts: &[SentPacket], stats: &mut Stats) {
+        let acked = self.count_pmtud_probes(acked_pkts);
+        if acked == 0 {
+            return;
+        }
+        // A probe was ACKed, confir, the new MTU and try to probe upwards further.
+        stats.pmtud_ack += acked;
+        self.mtu = self.search_table[self.probe_index];
+        qdebug!("PMTUD probe of size {} succeeded", self.mtu);
+        self.start_pmtud();
+    }
+
+    /// Stops the PMTUD process, setting the MTU to the largest successful probe size.
+    fn stop_pmtud(&mut self, idx: usize, now: Instant) {
+        self.probe_state = Probe::NotNeeded; // We don't need to send any more probes
+        self.probe_index = idx; // Index of the last successful probe
+        self.mtu = self.search_table[idx]; // Leading to this MTU
+        self.probe_count = 0; // Reset the count
+        self.loss_counts = vec![0; self.search_table.len()]; // Reset the loss counts
+        self.raise_timer = Some(now + PMTU_RAISE_TIMER);
+        qdebug!(
+            "PMTUD stopped, PLPMTU is now {}, raise timer {:?}",
+            self.mtu,
+            self.raise_timer.unwrap()
+        );
+    }
+
+    /// Checks whether a PMTUD probe has been lost. If it has been lost more than `MAX_PROBES`
+    /// times, the PMTUD process is stopped.
+    pub fn on_packets_lost(
+        &mut self,
+        lost_packets: &[SentPacket],
+        stats: &mut Stats,
+        now: Instant,
+    ) {
+        // Track lost probes
+        let lost = self.count_pmtud_probes(lost_packets);
```

Comment:
Is there a scenario where `lost` can be larger than `1`? In other words, is there a scenario where more than one probe is in-flight at once?

If I understand correctly `prepare_probe` is only called in `Probe::Needed`. `prepare_probe` switches to `Probe::Prepared`. The state only switches back to `Probe::Needed`, when the in-flight probe has been lost. Thus there is never more than one probe in-flight.

If the above is correct, why is there a `count_pmtud_probes` function? Am I missing something?

Reply:
@martinthomson does this implement what you suggested?

Reply:
> So maybe this whole thing isn't quite right and you only need to track whether the probes worked out (which is far simpler as well).

If we did that, we need a different mechanism to detect when a path change causes the PMTU to drop.